### PR TITLE
Proof of concept support for multiple rosdistro urls

### PR DESCRIPTION
**Pull request by tfoote**
This is a basic demo. There's at least one catch I'm surprised to need to do.
And this does not scale well. It takes a very long time to index and  scrape.
I wasn't able to get jekyll serve to work with just ardent, bouncy, and indigo being indexed.

The traceback for the todo is: https://gist.github.com/tfoote/9bfbe5201357567e92cbcde288ce0464

it took 950 seconds to scrape indigo, ardent and bouncy.

I've launched it in the background with these full settings but we're likely looking at an hour or more to run. And I don't see a way to test it besides deploying it.

Connects to #43

### CODE REVIEW

**Comment by tfoote** [comment]
Some timing a full rebuild was 90 minutes
```

Generating update report...
Found 4784 packages total.
Generating package pages...
Generating repo list pages...
Generating package list pages...
Generating rosdep list pages...
Generating search index...
Precompiling lunr index...
Full lunr cmd: /workdir/rosindex/node_modules/lunr-index-build/bin/lunr-index-build -r id -f baseurl -f instance -f url -f tags:100 -f name:100 -f version -f description:50 -f maintainers -f authors -f distro -f readme -f released -f unreleased < /workdir/rosindex/_remotes/deploy/search.json > /workdir/rosindex/_remotes/deploy/index.json
  indexing [===================] 100% 0.0s
Generating statistics page...
Generating errors page...
        Generating: Indexer finished in 4780.530110278 seconds.
        Generating: JekyllOptionalFrontMatter::Generator finished in 0.128112654 seconds.
        Generating: JekyllReadmeIndex::Generator finished in 3.552718012 seconds.
        Generating: Jekyll::JekyllSitemap finished in 0.002717659 seconds.
        Generating: JekyllTitlesFromHeadings::Generator finished in 0.036939682 seconds.
        Generating: JekyllRelativeLinks::Generator finished in 25.604016652 seconds.
...
                    done in 5369.918 seconds.
 Auto-regeneration: disabled. Use --watch to enable.

real	89m42.838s
user	35m59.096s
sys	13m55.196s
```

Other issues

The cache is too big for github
```
Counting objects: 42781, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (27769/27769), done.
Writing objects: 100% (42781/42781), 77.48 MiB | 726.00 KiB/s, done.
Total 42781 (delta 21036), reused 0 (delta 0)
remote: Resolving deltas: 100% (21036/21036), completed with 122 local objects.
remote: warning: File cache/_deploy_cache.db.bak is 67.00 MB; this is larger than GitHub's recommended maximum file size of 50.00 MB
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
remote: error: Trace: 42363def81fc3c8c1b650874d1399e0c
remote: error: See http://git.io/iEPt8g for more information.
remote: error: File cache/_deploy_cache.db is 222.93 MB; this exceeds GitHub's file size limit of 100.00 MB
To git@github.com:ros-infrastructure/index.ros.org.git
 ! [remote rejected] gh-pages -> gh-pages (pre-receive hook declined)
error: failed to push some refs to 'git@github.com:ros-infrastructure/index.ros.org.git'
root@snowman:/workdir/rosindex/_remotes/deploy# git branch
* gh-pages
root@snowman:/workdir/rosindex/_remotes/deploy# git status
On branch gh-pages
Your branch is ahead of 'origin/gh-pages' by 2 commits.
  (use "git push" to publish your local commits)
nothing to commit, working directory clean
root@snowman:/workdir/rosindex/_remotes/deploy# git rm cache/_deploy_
_deploy_cache.db                _deploy_cache.db.bak            _deploy_index_report.yaml       _deploy_index_report_diff.yaml
root@snowman:/workdir/rosindex/_remotes/deploy# git rm cache/_deploy_*
rm 'cache/_deploy_cache.db'
rm 'cache/_deploy_cache.db.bak'
rm 'cache/_deploy_index_report.yaml'
rm 'cache/_deploy_index_report_diff.yaml'
root@snowman:/workdir/rosindex/_remotes/deploy# git commit --amend
```

The repo is also getting large for github. We're at 1.5GB

**Comment by hidmic** [comment]
Hmm, 90 minutes is insane. I wonder what fraction of that time is spent fetching sources only (I/O over network, I/O to disk). About the cache, I guess we can pull it out of the deployment site and keep it local. We can play with docker volumes later on.

File: _config.yml [hunk 1]
```diff
@@ -79,9 +79,16 @@ index_old_doc_paths: false
...
   - 'bouncy'
+  - 'melodic'
+  - 'lunar'
+  - 'kinetic'
+  - 'indigo'
+
 old_distros:
   - 'ardent'
-
+  - 'jade'
```
Comment (hidmic) [inline _config.yml:89]: @tfoote just curious: why is `jade` an _old_ distro if `indigo` is a distro?

File: _plugins/rosindex_generator.rb [hunk 2]
```diff
@@ -831,225 +831,227 @@ def generate(site)
       $all_distros.reverse_each do |distro|
 
         puts "processing rosdistro: "+distro
+        # TODO(tfoote) For loop here
```
Comment (hidmic) [inline _plugins/rosindex_generator.rb]: @tfoote I don't fully grasp this TODO.

**Review by hidmic — APPROVED** [review]
LGTM

File: _config.yml (hunk 1 above)
Comment (tfoote) [reply _config.yml:89]: Jade was EOL in May 2017 Indigo is an LTS on Trusty so it's valid through April 2019

http://wiki.ros.org/Distributions

File: _plugins/rosindex_generator.rb (hunk 2 above)
Comment (tfoote) [reply _plugins/rosindex_generator.rb]: Sorry this todo was resolved I forgot to remove it

File: _config.yml (hunk 1 above)
Comment (hidmic) [reply _config.yml:89]: Ohh. Good point.

**Comment by hidmic** [comment]
Ready to merge @tfoote ! :)

**Comment by tfoote** [comment]
My estimate of timing is that it's about 55 minutes to do the downloads and 30 minutes to do the scraping and 5 minutes to do the generation.

**Comment by hidmic** [comment]
Hmm, there's not much we can do about I/O (besides caching repo clones) but 30 minutes to scrape is a long time. Maybe we're not starting enough workers? Maybe we can profile the rosindex generator plugin at some point.

**Pull request merged**